### PR TITLE
43: asserts, tests, and code cleaning

### LINKED ISSUE #43: Add assertions to any functions that make assumptions about input parameters
Example:
```
/*
 * Returns the sum of all corresponding elements from two matrices being
 * multiplied together. For example:
 *   a = 1 2   b = 4 1
 *       3 4       2 3
 *   multiplyCorrespondingElementsAndSum(a, b)
 *   (1*4) + (2*1) + (3*2) + (4*3) = 21
 *
 * m1: the first matrix
 * m2: the second matrix (of the same dimensions as m1)
 */
export function multiplyCorrespondingElementsAndSum(m1, m2) {
  let condition = (m1.length === m2.length);
  let errorMessage = 'multiplyCorrespondingElementsAndSum: m1 and m2 are not the same width';
  utils.assert(condition, errorMessage);

  condition = (m1.length === m2.length);
  errorMessage = 'multiplyCorrespondingElementsAndSum: m1 and m2 are not the same height';
  utils.assert(condition, errorMessage);

  const mWidth = m1.length;
  const mHeight = m1[0].length;
  let sum = 0;

  for (let x = 0; x < mWidth; x++) {
    for (let y = 0; y < mHeight; y++) {
      sum += m1[x][y] * m2[x][y];
    }
  }

  return sum;
}
```

### LINKED ISSUE #50: assert that input `cpt` is a divisor of 40 (1, 2, 4, 5, 8, 10, 20, or 40)

### PR DESCRIPTION
**Pull request by davidabrahams**
This is a LARGE code-improvement PR. There are no functional changes to the application.

TLDR: branch closes issues #43, #50

1. Functions that make assumptions about input parameters now have asserts, and corresponding unit tests that check that the function explode when given improper input. This branch closes #43 and closes #50 (both assertion issues).
2. Unit tests for `isTraversable` and `getShortestPath` (we can check A* is working!)
3. In order to do the A* test, I added the javascript-astar library as an NPM package, and updated dependencies accordingly. This also just generally seems smart, so that all our dependencies are handled in one way (rather than using the `include` tag tagpro provides)
4. Some small code improvement things, like throwing exceptions in functions when something unexpected happens instead of returning an input, so that we can catch bugs (this is actually how I found #78), and adding default values to some function arguments that felt like they should have them.

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -13,7 +13,6 @@
   "globals": {
     "tagpro": false,
     "Box2D": false,
-    "astar": false,
-    "Graph": false,
+    "PIXI": false,
```
Comment (billmwong) [inline .eslintrc]: Why is this here

File: src/helpers/map.js [hunk 2]
```diff
@@ -315,12 +318,16 @@ export function getSubarrayFrom2dArray(array, xCenter, yCenter, width, height) {
 export function convolve(m, k) {
   const kWidth = k.length;
   const kHeight = k[0].length;
+  // TODO: Why can't you apply rectangular kernals? I know why you wouldn't want to, but you should
```
Comment (shanek21) [inline src/helpers/map.js]: The answer is always that Shane was being lazy. I made an issue to fix this.

File: src/helpers/map.js [hunk 3]
```diff
@@ -59,7 +59,7 @@ export function isTraversable(tileID) {
     case tileTypes.BLUE_GATE:
       return amBlue();
     default:
-      return false;
+      throw new Error(`Unknown tileID given function: ${tileID}`);
```
Comment (davidabrahams) [inline src/helpers/map.js]: oh i think i meant to say "given to function" or something. yeah i'll take the "function" word out

**Comment by davidabrahams** [comment]
@shanek21 I think your convention is better for assert errors. I did some random python errors, and they seem to tell you exactly what you did wrong, instead of what you should have done. I say we stick to that:

```python
>>> print max(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not iterable
>>> print [1, 2, 3][4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```

good catch. i'll go through and fix that

File: src/helpers/map.js (hunk 2 above)
Comment (billmwong) [inline src/helpers/map.js]: If an issue exists, no need to document in a code comment here

File: src/helpers/map.js (hunk 3 above)
Comment (shanek21) [inline src/helpers/map.js]: You still didn't change this?

File: src/helpers/map.js (hunk 2 above)
Comment (shanek21) [inline src/helpers/map.js]: Or this hahaha

File: src/helpers/map.js [hunk 4]
```diff
@@ -322,6 +325,8 @@ export function convolve(m, k) {
   const mWidth = m.length;
   const mHeight = m[0].length;
   const kSize = kWidth;
+  assert(kSize <= mWidth && kSize <= mHeight,
+    'kernal size must be smaller than both matrix dimensions');
```
Comment (shanek21) [inline src/helpers/map.js]: Should be 'kernal size is not smaller than both matrix dimensions'

File: src/helpers/path.js [hunk 5]
```diff
@@ -54,3 +64,4 @@ export function getTarget(myX, myY, shortestPath) {
   const res = { x: next.x, y: next.y };
   return res;
 }
+
```
Comment (billmwong) [inline src/helpers/path.js]: Why

File: test/map.spec.js [hunk 6]
```diff
@@ -3,6 +3,21 @@ import * as map from '../src/helpers/map';
...
 
+test('test isTraversable', t => {
+  t.true(map.isTraversable(2)); // Regular floor
+  t.true(map.isTraversable(3.1)); // taken red flag
+  t.true(map.isTraversable(9)); // inactive gate
+  t.true(map.isTraversable(17)); // red endzone
+  t.false(map.isTraversable(0)); // Blank space
+  t.false(map.isTraversable(1)); // square wall
+  t.false(map.isTraversable(7)); // spike
+  // some inputs that the function shouldn't handle
```
Comment (billmwong) [inline test/map.spec.js]: Split this into two tests, like "returns correct traversable value" and "throws error for incorrect inputs"

File: test/map.spec.js [hunk 7]
```diff
@@ -42,7 +57,11 @@ test('test fillGridWithSubgrid', t => {
   ];
   map.fillGridWithSubgrid(grid, subgrid, 1, 0);
   t.same(grid, expected);
-
+  t.doesNotThrow(() => { map.fillGridWithSubgrid(grid, subgrid, 0, 0); });
```
Comment (billmwong) [inline test/map.spec.js]: I think we should do the same for all of these too. Although we might need to talk on slack about this decision. I think test names should be descriptive and specific so that failures are easier to track down

File: src/helpers/path.js (hunk 5 above)
Comment (davidabrahams) [inline src/helpers/path.js]: like this should be a one liner?

File: test/map.spec.js (hunk 6 above)
Comment (davidabrahams) [inline test/map.spec.js]: kk

File: test/map.spec.js [hunk 8]
```diff
@@ -3,23 +3,26 @@ import * as map from '../src/helpers/map';
...
   t.true(map.isTraversable(9)); // inactive gate
   t.true(map.isTraversable(17)); // red endzone
   t.false(map.isTraversable(0)); // Blank space
   t.false(map.isTraversable(1)); // square wall
   t.false(map.isTraversable(7)); // spike
-  // some inputs that the function shouldn't handle
+  t.end();
+});
+
+test('test isTraversable throws errors for inproper inputs', t => {
```
Comment (billmwong) [inline test/map.spec.js]: inproper -> improper

File: test/map.spec.js [hunk 9]
```diff
@@ -135,24 +187,34 @@ test('test traversableCellsInTile', t => {
...
+    [1, 0, 0, 0, 0, 0, 0, 1],
+    [1, 0, 0, 0, 0, 0, 0, 1],
+    [1, 0, 0, 0, 0, 0, 0, 1],
+    [1, 0, 0, 0, 0, 0, 0, 1],
+    [1, 1, 0, 0, 0, 0, 1, 1],
+    [1, 1, 1, 1, 1, 1, 1, 1],
+  ];
+  t.same(map.traversableCellsInTile(false, 8, 14), expected);
+
+  t.throws(() => { map.traversableCellsInTile(false, 7, 14); });
```
Comment (billmwong) [inline test/map.spec.js]: separate out here

File: test/map.spec.js [hunk 10]
```diff
@@ -3,8 +3,26 @@ import * as map from '../src/helpers/map';
...
+  t.true(map.isTraversable(3.1)); // taken red flag
+  t.true(map.isTraversable(9)); // inactive gate
+  t.true(map.isTraversable(17)); // red endzone
+  t.false(map.isTraversable(0)); // Blank space
+  t.false(map.isTraversable(1)); // square wall
+  t.false(map.isTraversable(7)); // spike
+  t.end();
+});
+
+test('test isTraversable throws errors for inproper inputs', t => {
```
Comment (billmwong) [inline test/map.spec.js]: use consistent language, improper or invalid?

File: test/path.spec.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    [1, 1, 1, 1, 1, 1, 1, 1],
+    [1, 1, 1, 1, 1, 1, 1, 1],
+    [0, 0, 0, 0, 0, 0, 0, 0],
+    [1, 1, 1, 1, 1, 1, 1, 1],
+    [1, 1, 1, 1, 1, 1, 1, 1],
+  ];
+  const impossiblePath = getShortestPath(0, 4, 5, 5, impossibleMap, false);
+  t.notOk(impossiblePath);
+
+  t.throws(() => { getShortestPath(6, 3, 3, 6, impossibleMap, false); });
```
Comment (billmwong) [inline test/path.spec.js]: separate this out

File: test/path.spec.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import test from 'tape';
+import { getShortestPath } from '../src/helpers/path';
+
+test('test getShortestPath', t => {
```
Comment (billmwong) [inline test/path.spec.js]: make the name of this test "test getShortestPath" more specific

File: src/helpers/path.js [hunk 13]
```diff
@@ -51,6 +61,6 @@ export function getTarget(myX, myY, shortestPath) {
   }
   const next = shortestPath[j];
   // TODO: this seems to throw null pointer when bot doesn't know where the center flag is.
-  const res = { x: next.x, y: next.y };
-  return res;
+  return { x: next.x, y: next.y };
 }
+
```
Comment (billmwong) [inline src/helpers/path.js]: get rid of this line

File: test/map.spec.js (hunk 7 above)
Comment (davidabrahams) [inline test/map.spec.js]: done

File: test/map.spec.js (hunk 8 above)
Comment (davidabrahams) [inline test/map.spec.js]: done

File: test/map.spec.js (hunk 10 above)
Comment (davidabrahams) [inline test/map.spec.js]: done

File: test/path.spec.js (hunk 11 above)
Comment (davidabrahams) [inline test/path.spec.js]: done

File: test/path.spec.js (hunk 12 above)
Comment (davidabrahams) [inline test/path.spec.js]: done

File: src/helpers/path.js (hunk 13 above)
Comment (davidabrahams) [inline src/helpers/path.js]: done

File: src/helpers/path.js [hunk 14]
```diff
@@ -51,6 +61,5 @@ export function getTarget(myX, myY, shortestPath) {
   }
   const next = shortestPath[j];
   // TODO: this seems to throw null pointer when bot doesn't know where the center flag is.
```
Comment (billmwong) [inline src/helpers/path.js]: This todo was because of that tile types error right? Which is resolved

Comment (davidabrahams) [inline src/helpers/path.js]: done nice catch

File: src/helpers/path.js [hunk 15]
```diff
@@ -9,16 +12,23 @@
...
+
   const graph = new Graph(grid, { diagonal });
   const start = graph.grid[myX][myY];
   const end = graph.grid[targetX][targetY];
   // calculate shortest path list
   const shortestPath = astar.search(graph, start, end,
     { heuristic: diagonal ? astar.heuristics.diagonal : astar.heuristics.manhattan });
+
+  if (shortestPath.length === 0) {
+    return undefined;
```
Comment (altodyte) [inline src/helpers/path.js]: Is this just to make the return more explicit? This feels like a weird way to handle this, though given the removal of the TODO I imagine that it fixes the issue arising during the 0.25s when the flag doesn't exist. Could you elaborate why this fixes that?

Comment (davidabrahams) [inline src/helpers/path.js]: that problem is fixed, this is not origin of that fix. see PR #81 for how that problem was fixed.

The reason we return undefined here is so that if no shortest path is calculated, the code will null pointer elsewhere, letting us know something went wrong.
